packages/matter.js/src/codec/BtpCodec.ts

```diff
@@ -0,0 +1,151 @@
+        const reader = new DataReader(data, Endian.Little);
+
+        return {
+            header: this.decodeHeader(reader),
+            payload: this.decodeBtpPacketPayload(reader)
+        };
+
+    }
+
+    static encodeBtpHandshakeResponse({ version, attMtu, windowSize }: BtpHandshakeResponse): ByteArray {
+
+        return ByteArray.concat(
+            new Uint8Array(HANDSHAKE_HEADER),
+            this.encodeHandshakeResponsePayload({ version, attMtu, windowSize })
+        );
+    }
+
+    private static decodeBtpPacketPayload(reader: DataReader<Endian.Little>): BtpPacketPayload {
+
+        const ackNumber = reader.readUInt8();
+        const sequenceNumber = reader.readUInt8();
+        const msgLength = reader.readUInt16();
+        const segmentPayload = reader.getRemainingBytes();
+
+        return { ackNumber, sequenceNumber, msgLength, segmentPayload };
+    }
+
+
+    private static decodeRequestPayload(reader: DataReader<Endian.Little>): BtpHandshakeRequest {
+
+        const header = reader.readUInt8();
+        if (header !== HANDSHAKE_HEADER) throw new Error("Handshake Error - Incorrect BTP Handhshake Request Headers");
+
+        const opcode = reader.readUInt8();
+        if (opcode !== Opcode.HandshakeManagementOpcode) throw new Error("Handshake Error - Management Opcode is incorrect");
+
+        var versions = reader.readInt8();
+        ver[0] = versions & 0xF0;
+        ver[1] = versions & 0x0F;
+
+        versions = reader.readInt8();
+        ver[2] = versions & 0xF0;
+        ver[3] = versions & 0x0F;
+
+        versions = reader.readInt8();
+        ver[4] = versions & 0xF0;
+        ver[5] = versions & 0x0F;
+
+        versions = reader.readInt8();
+        ver[6] = versions & 0xF0;
+        ver[7] = versions & 0x0F;
+
+        const version = ver.filter(v => v !== 0)
+        if(version.length == 0) throw new Error("Handshake Error - Version is incorrect");
+
+        const attMtu = reader.readUInt16();
+        const clientWindowSize = reader.readUInt8();
+
+        return { version, attMtu, clientWindowSize };
+    }
+
+    private static decodeHeader(reader: DataReader<Endian.Little>): Header {
+
+        const headerBits = reader.readUInt8();
+        const handshakeBit = (headerBits & HeaderBits.HandshakeBit);
+        const managementBit = (headerBits & HeaderBits.ManagementMsg);
+        const ackMsgBit = (headerBits & HeaderBits.AckMsg);
+        const endSegmentBit = (headerBits & HeaderBits.EndSegment);
+        const beginSegmentBit = (headerBits & HeaderBits.BeginSegment);
+
+        return { handshakeBit, managementBit, ackMsgBit, endSegmentBit, beginSegmentBit };
+    }
+
+    private static encodeHandshakeResponsePayload({ version, attMtu, windowSize }: BtpHandshakeResponse): ByteArray {
+
+        const writer = new DataWriter(Endian.Little);
+
+        writer.writeUInt8(Opcode.HandshakeManagementOpcode);
+        const maxVersion = Math.max(...version); // finding out the highest version
```

Comment:
I would say here it just gets "one version number" and nt the array that casme in. The BTPSessionHandler will have te logic to deide for the version. Here it is just "dump encoding"

Reply:
Updated